packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -1,10 +1,164 @@
-import { injectable, W3cCredentialService } from '@aries-framework/core'
+import type { AgentContext, Logger, W3cCredentialRecord } from '@aries-framework/core'
+import type { Jwt } from '@sphereon/openid4vci-client'
+
+import {
+  AgentConfig,
+  AriesFrameworkError,
+  injectable,
+  JsonEncoder,
+  JsonTransformer,
+  W3cCredentialService,
+  W3cVerifiableCredential,
+} from '@aries-framework/core'
+import {
+  Alg,
+  AuthzFlowType,
+  CredentialRequestClientBuilder,
+  OpenID4VCIClient,
+  ProofOfPossessionBuilder,
+} from '@sphereon/openid4vci-client'
+
+import { JwsService } from '../../core/src/crypto/JwsService'
+import { didKeyToVerkey } from '../../core/src/modules/dids/helpers'
+
+export interface PreAuthorizedOptions {
+  issuerUri: string
+  kid: string
+}
 
 @injectable()
 export class OpenId4VcClientService {
+  private logger: Logger
   private w3cCredentialService: W3cCredentialService
+  private jwsService: JwsService
 
-  public constructor(w3cCredentialService: W3cCredentialService) {
+  public constructor(agentConfig: AgentConfig, w3cCredentialService: W3cCredentialService, jwsService: JwsService) {
     this.w3cCredentialService = w3cCredentialService
+    this.jwsService = jwsService
+    this.logger = agentConfig.logger
+  }
+
+  private signCallback(agentContext: AgentContext) {
+    return async (jwt: Jwt, kid: string) => {
+      // TODO should we check if the did exists here, or just let the wallet throw?
+
+      if (!jwt.header) {
+        throw new AriesFrameworkError('No header present on JWT')
+      }
+
+      if (!jwt.payload) {
+        throw new AriesFrameworkError('No payload present on JWT')
+      }
+
+      const did = kid.split('#')[0]
+
+      const verkey = didKeyToVerkey(did)
```

Comment:
I had to add the `did` property as an optional propety to getCreatedDids in the DidsApi and `DidsRepository`

Reply:
Thanks! Adding this as we speak. The OpenId4VCI spec doesn't specify what the proof purpose should be. Since this is about proving control over a certain DID, I'm assuming your suggestion here is right..